olmo/scaling/scaling_laws/extrapolate_n.py

```diff
@@ -0,0 +1,70 @@
+import csv
+from dataclasses import dataclass
+from typing import List, Dict
+
+import matplotlib.pyplot as plt
+import numpy as np
+import scipy
+
+from .curve_fit import get_coefficients
+
+
+@dataclass
+class CurveFitConfig:
+    path: str
+    """
+    Path containing the W&B downloaded data and metadata.
+    """
+
+    keys: List[str]
+    """
+    The metrics for computing the scaling law predictions.
+    """
+
+    n: int
+    """
+    The number of parameters in the model.
+    """
+
+    mode: str
+    """
+    Whether this model is used for fitting the curve ('train') or evaluating the fit ('eval').
+    """
+
+
+def get_data_at_d(config_by_n: Dict[int, CurveFitConfig], d: int):
+    """
+    d: If its value is string "last", then loss from the last ckpt is used.
+       If its value is an integer, then loss from the first ckpt with at least d tokens is used.
+    """
+    train_ns, train_ys, eval_ns, eval_ys = [], [], [], []
+    for n, config in config_by_n.items():
+        with open(config.path) as file_ref:
+            reader = csv.DictReader(file_ref)
+            y = None
+            for row in reader:
+                dd = int(float(row['throughput/total_tokens']))
+                yy = np.mean([float(row[key]) for key in config.keys])
+                if dd >= d and y is None:
+                    y = yy
+            if y is None: # there is no data at or later than d tokens
+                continue
+            if config.mode == 'train':
+                train_ns.append(n)
+                train_ys.append(y)
+            elif config.mode == 'eval':
+                eval_ns.append(n)
+                eval_ys.append(y)
+    return train_ns, train_ys, eval_ns, eval_ys
```

Comment:
Isn't this a pared-down version of what's in the `curve_fit.py` file already? Can we combine these two files?

Reply:
Details are a bit different. I feel it'd be more work to consolidate them than keeping them separate.

Plus I think we should rename `curve_fit.py` to `extrapolate_d.py` to distinguish it from extrapolate_n ...